tests/util/bft_network_partitioning.py

```diff
@@ -0,0 +1,116 @@
+# You may not use this product except in compliance with the Apache 2.0 License.
+#
+# This product may include a number of subcomponents with separate copyright
+# notices and license terms. Your use of these subcomponents is subject to the
+# terms and conditions of the subcomponent's license, as noted in the LICENSE
+# file.
+import subprocess
+from abc import ABC, abstractmethod
+from itertools import combinations
+
+
+class NetworkPartitioningAdversary(ABC):
+    """Represents an adversary capable of inflicting network partitioning"""
+
+    BFT_NETWORK_PARTITIONING_RULE_CHAIN = "bft-network-partitioning"
+
+    def __init__(self, bft_network):
+        self.bft_network = bft_network
+
+    def __enter__(self):
+        """context manager method for 'with' statements"""
+        self._init_bft_network_rule_chain()
+
+        return self
+
+    def __exit__(self, *args):
+        """context manager method for 'with' statements"""
+        self._remove_bft_network_rule_chain()
+
+    @abstractmethod
+    def interfere(self):
+        """ This is where the actual malicious behavior is defined """
+        pass
+
+    @classmethod
+    def _init_bft_network_rule_chain(cls):
+        subprocess.run(
+            ["iptables", "-N", cls.BFT_NETWORK_PARTITIONING_RULE_CHAIN],
+            check=True)
+        subprocess.run(
+            ["iptables", "-A", "INPUT",
+             "-s", "localhost", "-d", "localhost",
+             "-j", cls.BFT_NETWORK_PARTITIONING_RULE_CHAIN],
+            check=True)
+
+    @classmethod
+    def _remove_bft_network_rule_chain(cls):
+        subprocess.run(
+            ["iptables", "-D", "INPUT",
+             "-s", "localhost", "-d", "localhost",
+             "-j", cls.BFT_NETWORK_PARTITIONING_RULE_CHAIN],
+            check=True)
+        subprocess.run(
+            ["iptables", "-F", cls.BFT_NETWORK_PARTITIONING_RULE_CHAIN], check=True)
+        subprocess.run(
+            ["iptables", "-X", cls.BFT_NETWORK_PARTITIONING_RULE_CHAIN], check=True)
+
+    @classmethod
+    def _drop_packets_between(
+            cls, source_port, dest_port, drop_rate_percentage=100):
+        assert 0 <= drop_rate_percentage <= 100
+        drop_rate = drop_rate_percentage / 100
+        subprocess.run(
+            ["iptables", "-A", cls.BFT_NETWORK_PARTITIONING_RULE_CHAIN,
+             "-p", "udp",
+             "--sport", str(source_port), "--dport", str(dest_port),
+             "-m", "statistic", "--mode", "random",
+             "--probability", str(drop_rate),
+             "-j", "DROP"],
+            check=True
+        )
+
+
+class PassiveAdversary(NetworkPartitioningAdversary):
+    """ Adversary does nothing = synchronous network """
+
+    def interfere(self):
+        pass
+
+
+class PrimaryIsolatingAdversary(NetworkPartitioningAdversary):
+    """ Adversary that intercepts and drops all outgoing packets from the current primary """
+
+    async def interfere(self):
```

Comment:
Does overriding a def abstract method with async def make a difference?

Reply:
It doesn't seem to... Initially I had made the abstract method async, but then I thought - do we want this to be necassarily a coroutine? Probably not. Such as I see it, "async" is more of an implementation detail, so I moved it to the concrete classes (where applicable).